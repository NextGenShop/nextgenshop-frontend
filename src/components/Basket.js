import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import DeleteIcon from "@material-ui/icons/Delete";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import Avatar from "@material-ui/core/Avatar";
import IconButton from "@material-ui/core/IconButton";
import { Typography } from "@material-ui/core";
import PlaceholderImage from "../assets/images/placeholder_image.png";
import { connect } from "react-redux";
import { actions } from "../store/api/basket";
import { removeBasketItem } from "../utils/basketUtils";
import Button from "@material-ui/core/Button";
import { Link } from "react-router-dom";

const useStyles = makeStyles((theme) => ({
  list: {
    height: "29vh",
    overflowY: "scroll",
  },
  price: {
    color: theme.palette.primary.dark,
    fontWeight: 600,
    marginRight: theme.spacing(2),
  },
  total: {
    fontWeight: "bold",
  },
}));

export function Basket({
  shopperId,
  dispatchUpdateBasket,
  dispatchGetBasket,
  basket,
}) {
  const classes = useStyles();

  React.useEffect(() => {
    dispatchGetBasket(shopperId);
  }, [dispatchGetBasket, shopperId]);

  const removeItem = (productId) => {
    const newBasket = removeBasketItem(productId, basket);
    dispatchUpdateBasket(shopperId, newBasket);
  };

  return (
    <React.Fragment>
      <List className={classes.list}>
        {basket && basket.items && basket.items.length > 0 ? (
          basket.items.map((item) => (
            <ListItem key={item.product.productId}>
              <ListItemAvatar>
                <Avatar
                  src={
                    item.product.image === "mockImageUrl"
                      ? PlaceholderImage
                      : item.product.image
                  }
                />
              </ListItemAvatar>
              <ListItemText
                primary={item.product.name}
                secondary={"Qty: " + item.quantity}
              />
              <ListItemSecondaryAction>
                <Typography className={classes.price} display="inline">
                  £{(item.product.price * item.quantity).toFixed(2)}
                </Typography>
                <IconButton
                  edge="end"
                  aria-label="delete"
                  onClick={() => removeItem(item.product.productId)}
                >
                  <DeleteIcon />
                </IconButton>
              </ListItemSecondaryAction>
            </ListItem>
          ))
        ) : (
          <Typography variant="subtitle2">
            Your shopping basket is empty.
          </Typography>
        )}
      </List>
      <Typography className={classes.total} variant="h6" component="div">
        Total:{" "}
        <Typography variant="h6" display="inline">
          £ {basket && basket.totalPrice ? basket.totalPrice.toFixed(2) : 0}
        </Typography>
      </Typography>
      <Link to="/checkout"> Checkout </Link>
    </React.Fragment>
  );
}

const mapStateToProps = (state) => ({
  basket: state.basket,
});

const mapDispatchToProps = (dispatch) => ({
  dispatchGetBasket: (shopperId) => dispatch(actions.getBasket(shopperId)),
  dispatchUpdateBasket: (shopperId, basketData) =>
    dispatch(actions.updateBasket(shopperId, basketData)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Basket);
