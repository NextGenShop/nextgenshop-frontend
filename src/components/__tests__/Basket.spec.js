import React from 'react';
import { shallow } from 'enzyme';
import { Basket } from '../Basket';

describe('Basket', () => {
  test('should correctly render', () => {
    const component = shallow(<Basket />);
    expect(component).toMatchSnapshot();
  });
  test('should correctly render with props', () => {
    const component = shallow(<Basket shopperId='mockId' />);
    expect(component).toMatchSnapshot();
  });
});
