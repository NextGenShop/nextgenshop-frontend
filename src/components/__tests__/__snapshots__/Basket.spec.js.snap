// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket should correctly render 1`] = `
<Fragment>
  <WithStyles(ForwardRef(List))
    className="makeStyles-list-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="subtitle2"
    >
      Your shopping basket is empty.
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-total-3"
    component="div"
    variant="h6"
  >
    Total:
     
    <WithStyles(ForwardRef(Typography))
      display="inline"
      variant="h6"
    >
      £ 
      0
    </WithStyles(ForwardRef(Typography))>
    <Link
      className="makeStyles-checkoutLink-4"
      to="/checkout"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        variant="outlined"
      >
        Checkout
      </WithStyles(ForwardRef(Button))>
    </Link>
  </WithStyles(ForwardRef(Typography))>
</Fragment>
`;

exports[`Basket should correctly render a single basket item 1`] = `
<Fragment>
  <WithStyles(ForwardRef(List))
    className="makeStyles-list-1"
  >
    <WithStyles(ForwardRef(ListItem))
      key="0"
    >
      <WithStyles(ForwardRef(ListItemAvatar))>
        <WithStyles(ForwardRef(Avatar))
          src="placeholder_image.png"
        />
      </WithStyles(ForwardRef(ListItemAvatar))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Milk Chocolate Biscuits 9 Pack"
        secondary="Qty: 1"
      />
      <WithStyles(ForwardRef(ListItemSecondaryAction))>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-price-2"
          display="inline"
        >
          £
          1.50
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="deleteButton"
          edge="end"
          onClick={[Function]}
        >
          <Memo([object Object]) />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(ListItemSecondaryAction))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-total-3"
    component="div"
    variant="h6"
  >
    Total:
     
    <WithStyles(ForwardRef(Typography))
      display="inline"
      variant="h6"
    >
      £ 
      0
    </WithStyles(ForwardRef(Typography))>
    <Link
      className="makeStyles-checkoutLink-4"
      to="/checkout"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        variant="outlined"
      >
        Checkout
      </WithStyles(ForwardRef(Button))>
    </Link>
  </WithStyles(ForwardRef(Typography))>
</Fragment>
`;

exports[`Basket should correctly render multiple basket items 1`] = `
<Fragment>
  <WithStyles(ForwardRef(List))
    className="makeStyles-list-1"
  >
    <WithStyles(ForwardRef(ListItem))
      key="0"
    >
      <WithStyles(ForwardRef(ListItemAvatar))>
        <WithStyles(ForwardRef(Avatar))
          src="placeholder_image.png"
        />
      </WithStyles(ForwardRef(ListItemAvatar))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Milk Chocolate Biscuits 9 Pack"
        secondary="Qty: 1"
      />
      <WithStyles(ForwardRef(ListItemSecondaryAction))>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-price-2"
          display="inline"
        >
          £
          1.50
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="deleteButton"
          edge="end"
          onClick={[Function]}
        >
          <Memo([object Object]) />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(ListItemSecondaryAction))>
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      key="1"
    >
      <WithStyles(ForwardRef(ListItemAvatar))>
        <WithStyles(ForwardRef(Avatar))
          src="placeholder_image.png"
        />
      </WithStyles(ForwardRef(ListItemAvatar))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Cookie Dough Vanilla Ice Cream 465Ml"
        secondary="Qty: 1"
      />
      <WithStyles(ForwardRef(ListItemSecondaryAction))>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-price-2"
          display="inline"
        >
          £
          4.50
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="deleteButton"
          edge="end"
          onClick={[Function]}
        >
          <Memo([object Object]) />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(ListItemSecondaryAction))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-total-3"
    component="div"
    variant="h6"
  >
    Total:
     
    <WithStyles(ForwardRef(Typography))
      display="inline"
      variant="h6"
    >
      £ 
      0
    </WithStyles(ForwardRef(Typography))>
    <Link
      className="makeStyles-checkoutLink-4"
      to="/checkout"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        variant="outlined"
      >
        Checkout
      </WithStyles(ForwardRef(Button))>
    </Link>
  </WithStyles(ForwardRef(Typography))>
</Fragment>
`;

exports[`Basket should correctly render with props 1`] = `
<Fragment>
  <WithStyles(ForwardRef(List))
    className="makeStyles-list-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="subtitle2"
    >
      Your shopping basket is empty.
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Typography))
    className="makeStyles-total-3"
    component="div"
    variant="h6"
  >
    Total:
     
    <WithStyles(ForwardRef(Typography))
      display="inline"
      variant="h6"
    >
      £ 
      0
    </WithStyles(ForwardRef(Typography))>
    <Link
      className="makeStyles-checkoutLink-4"
      to="/checkout"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        variant="outlined"
      >
        Checkout
      </WithStyles(ForwardRef(Button))>
    </Link>
  </WithStyles(ForwardRef(Typography))>
</Fragment>
`;
